### PR TITLE
Analyzer error handling

### PR DESCRIPTION
**Pull request by mnonnenmacher**
See the commit messages for details.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/56)

### CODE REVIEW

**Comment by sschuberth** [comment]
Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[analyzer/src/main/kotlin/managers/Maven.kt, line 23 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/56#-Kzh1NTzBBBBGkJIxCWL:-Kzh1NTzBBBBGkJIxCWM:b9420it) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/806f8b73e54de4e73ade61bf201b8f8090eeb314/analyzer/src/main/kotlin/managers/Maven.kt#L23)):*
> ```Kotlin
> 
> import ch.frankel.slf4k.*
> import com.here.ort.analyzer.Main
> ```

Nit: Put into block below.

---

*[model/src/main/kotlin/AnalyzerResult.kt, line 56 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/56#-Kzh3YKSYz29a1mBUW1V:-Kzh3YKSYz29a1mBUW1W:b-sbw95r) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/fc13fb856cfef90384b38ce3b26ca325c2043672/model/src/main/kotlin/AnalyzerResult.kt#L56)):*
> ```Kotlin
>      */
>     fun hasErrors(): Boolean {
>         fun hasErrors(pkgReference: PackageReference): Boolean =
> ```

Nit: Omit return type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/56)*

**Comment by sschuberth** [comment]
Reviewed 1 of 3 files at r3, 1 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/56)*

**Comment by sschuberth** [comment]
Reviewed 2 of 2 files at r6, 5 of 5 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/56)*

**Comment by sschuberth** [comment]
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[analyzer/src/funTest/kotlin/NpmTest.kt, line 62 at r9](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/56#-KziOlDJuUa4raJCmhwf:-KziOlDJuUa4raJCmhwg:b-unasr1) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/523e624e66b9033ca665fe037880ca632edba2ae/analyzer/src/funTest/kotlin/NpmTest.kt#L62)):*
> ```Kotlin
>             File(projectDir.parentFile, expectedResultsFilename).readText().let {
>                 if (OS.isWindows) {
>                     it.replace("/", "\\\\")
> ```

Given that we only need to do this for paths as part of exception error messages, we should probably at least add a comment about that fact, or maybe even add "Exception" to the search pattern.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/56)*

**Comment by sschuberth** [comment]
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/56)*

File: analyzer/src/funTest/kotlin/NpmTest.kt [hunk 1]
```diff
@@ -94,28 +92,31 @@ class NpmTest : WordSpec() {
...
-                }
+                val result = NPM.create().resolveDependencies(projectDir, listOf(packageFile))[packageFile]
 
-                @Suppress("UnsafeCallOnNullableType")
-                exception.message!! should endWith("contains multiple lockfiles. It is ambiguous which one to use.")
+                result shouldNotBe null
+                result!!.project shouldBe Project.createEmpty()
+                result.packages.size shouldBe 0
+                result.errors.size shouldBe 1
+                result.errors.first() shouldStartWith "IllegalArgumentException:"
```
Comment (sschuberth) [inline analyzer/src/funTest/kotlin/NpmTest.kt]: Can't we just use `should startWith("prefix")` / `should endWith("suffix")` (see [here](https://github.com/kotlintest/kotlintest/blob/master/doc/matchers.md))?

File: model/src/main/kotlin/AnalyzerResult.kt [hunk 2]
```diff
@@ -48,4 +48,14 @@ data class AnalyzerResult(
          * The list of errors that occurred during dependency resolution. Defaults to an empty list.
          */
         val errors: List<String> = emptyList()
-)
+) {
+    /**
+     * Return true if there were any errors during the analysis.
```
Comment (sschuberth) [inline model/src/main/kotlin/AnalyzerResult.kt]: ", false otherwise" ?

File: analyzer/src/funTest/kotlin/NpmTest.kt (hunk 1 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/kotlin/NpmTest.kt]: Done, had to switch to FreeSpec for this.

File: model/src/main/kotlin/AnalyzerResult.kt (hunk 2 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/AnalyzerResult.kt]: Done.

**Comment by mnonnenmacher** [comment]
Overriding approval for this PR because all review comments have been addressed an it contains important test fixes.

**Pull request merged**